index.js

```diff
@@ -196,14 +196,28 @@
     return _funcPath(false, path, implementations);
   }
 
+  /* istanbul ignore next */
+  //  functionName :: Function -> Nullable String
+  var functionName = 'name' in function f() {} ?
+      function functionName(f) { return f.name; } :
+      function functionName(f) {
+        var match = /function (\w*)/.exec(f);
+        return match && match[1];
+      };
+
   //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
   function $(_name, dependencies, requirements) {
     function getBoundMethod(_name) {
       var name = 'fantasy-land/' + _name;
       return requirements[_name] === Constructor ?
         function(typeRep) {
-          return funcPath([name], typeRep) ||
-                 implPath([/function (\w*)/.exec(typeRep)[1], name]);
+          var f = funcPath([name], typeRep);
+          if (f == null && typeof typeRep === 'function') {
+            var typeIdent = functionName(typeRep);
+            /* istanbul ignore else */
+            if (typeIdent) f = implPath([typeIdent, name]);
```

Comment:
I'd like to ensure that the condition evaluates to a Boolean value. What about the following?

```diff
-  //  functionName :: Function -> Nullable String
+  //  functionName :: Function -> String
```

```diff
-            if (typeIdent) f = implPath([typeIdent, name]);
+            if (typeIdent !== '') f = implPath([typeIdent, name]);
```